core/src/font.rs

```diff
@@ -405,38 +409,35 @@ impl<'gc> Font<'gc> {
 
 #[derive(Debug, Clone)]
 pub struct Glyph {
-    pub advance: i16,
     // Handle to registered shape.
     // If None, it'll be loaded lazily on first render of this glyph.
     shape_handle: Cell<Option<ShapeHandle>>,
+
     // Same shape as one in swf_glyph, but wrapped in an swf::Shape;
     // For use in hit tests. Created lazily on first use.
     // (todo: refactor hit tests to not require this?
     // this literally copies the shape_record, which is wasteful...)
     shape: RefCell<Option<swf::Shape>>,
+
     // The underlying glyph record, containing its shape.
     swf_glyph: swf::Glyph,
 }
 
 impl Glyph {
-    pub fn shape_handle(&self, renderer: &mut dyn RenderBackend) -> ShapeHandle {
-        if self.shape_handle.get().is_none() {
-            self.shape_handle
-                .set(Some(renderer.register_glyph_shape(&self.swf_glyph)))
-        }
-        self.shape_handle.get().unwrap()
+    pub fn as_shape(&self) -> Ref<'_, swf::Shape> {
+        self.shape
+            .borrow_mut()
+            .get_or_insert_with(|| ruffle_render::shape_utils::swf_glyph_to_shape(&self.swf_glyph));
+        Ref::map(self.shape.borrow(), |s| s.as_ref().unwrap())
     }
 
-    pub fn as_shape(&self) -> Ref<'_, swf::Shape> {
-        let mut write = self.shape.borrow_mut();
-        if write.is_none() {
-            *write = Some(ruffle_render::shape_utils::swf_glyph_to_shape(
-                &self.swf_glyph,
-            ));
-        }
-        drop(write);
-        let read = self.shape.borrow();
-        Ref::map(read, |s| s.as_ref().unwrap())
+    pub fn shape_handle(&self, renderer: &mut dyn RenderBackend) -> ShapeHandle {
+        self.shape_handle.get().unwrap_or_else(|| {
+            let shape_handle =
+                renderer.register_shape((&*self.as_shape()).into(), &NullBitmapSource);
```

Comment:
just for the record, copied from Discord:
> definitely don't want shape_handle to call as_shape
> check out the comment at shape field
> not sure if that part of impl Glyph refactor is needed at all

I think we settled on idea that maybe another refactor to consume the swf::Glyph would avoid doubling memory usage.

Reply:
Just rebased this PR on top of current master. Can we merge it as-is and implement said refactor in a separate PR?